File: olnatura-qr/api/olnaturaqr/src/main/java/com/company/olnaturaqr/support/label/EnvaseRestos.java
package com.company.olnaturaqr.support.label;

import com.company.olnaturaqr.domain.qr.QrLabel;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.web.server.ResponseStatusException;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.springframework.http.HttpStatus.BAD_REQUEST;

public final class EnvaseRestos {

    public static final int MAX = 4;
    private static final ObjectMapper JSON = new ObjectMapper();

    private EnvaseRestos() {}

    public static void apply(
            QrLabel q,
            List<String> cantidades,
            String cantidadPorEnvase,
            int envaseTotal
    ) {
        if (q == null) {
            throw new ResponseStatusException(BAD_REQUEST, "Etiqueta requerida");
        }
        List<String> cleaned = normalize(cantidades);
        if (cleaned.isEmpty()) {
            q.setRestosEnabled(false);
            q.setCantidadResto(null);
            q.setRestosCantidadesJson(null);
            return;
        }
        if (cleaned.size() > MAX) {
            throw new ResponseStatusException(BAD_REQUEST, "Máximo " + MAX + " etiquetas de resto");
        }
        if (envaseTotal < cleaned.size()) {
            throw new ResponseStatusException(BAD_REQUEST,
                    "La cantidad total de envases debe ser al menos " + cleaned.size());
        }
        String stdRaw = trimToNull(cantidadPorEnvase);
        if (stdRaw == null) {
            throw new ResponseStatusException(BAD_REQUEST, "Cantidad por envase es requerida");
        }
        BigDecimal std = parsePositive(stdRaw, "Cantidad por envase");
        for (int i = 0; i < cleaned.size(); i++) {
            String raw = cleaned.get(i);
            BigDecimal qty = parsePositive(raw, "Etiqueta de resto " + (i + 1));
            if (qty.compareTo(std) >= 0) {
                throw new ResponseStatusException(BAD_REQUEST,
                        "La etiqueta de resto " + (i + 1) + " debe ser menor a la cantidad por envase");
            }
        }
        q.setRestosEnabled(true);
        q.setRestosCantidadesJson(writeJson(cleaned));
        q.setCantidadResto(cleaned.get(cleaned.size() - 1));
    }

    public static List<String> listOf(QrLabel q) {
        if (q == null) return List.of();
        List<String> fromJson = readJson(q.getRestosCantidadesJson());
        if (!fromJson.isEmpty()) return fromJson;
        if (q.isRestosEnabled()) {
            String legacy = trimToNull(q.getCantidadResto());
            if (legacy != null) return List.of(legacy);
        }
        return List.of();
    }

    public static boolean isCantidadMenorEnvase(QrLabel q, int envaseNum) {
        List<String> list = listOf(q);
        if (list.isEmpty() || q == null) return false;
        int total = q.getEnvaseTotal();
        int first = total - list.size() + 1;
        return envaseNum >= first && envaseNum <= total;
    }

    public static String cantidadForEnvase(QrLabel q, int envaseNum) {
        if (q == null) return "N/A";
        List<String> list = listOf(q);
        if (!list.isEmpty()) {
            int total = q.getEnvaseTotal();
            int first = total - list.size() + 1;
            if (envaseNum >= first && envaseNum <= total) {
                return list.get(envaseNum - first);
            }
        }
        String std = trimToNull(q.getCantidadPorEnvase());
        return std == null ? "N/A" : std;
    }

    public static List<String> quantities(QrLabel q) {
        if (q == null) return List.of();
        int total = Math.max(0, q.getEnvaseTotal());
        List<String> out = new ArrayList<>(total);
        for (int i = 1; i <= total; i++) {
            out.add(cantidadForEnvase(q, i));
        }
        return Collections.unmodifiableList(out);
    }

    public static String cantidadTotal(QrLabel q) {
        List<String> qs = quantities(q);
        if (qs.isEmpty()) return "0";
        BigDecimal sum = BigDecimal.ZERO;
        int parsed = 0;
        for (String raw : qs) {
            String t = trimToNull(raw);
            if (t == null || "N/A".equalsIgnoreCase(t)) continue;
            try {
                sum = sum.add(parsePositive(t, "cantidad"));
                parsed++;
            } catch (RuntimeException ignored) {
            }
        }
        if (parsed == 0) return String.valueOf(qs.size());
        return sum.stripTrailingZeros().toPlainString();
    }

    static List<String> normalize(List<String> raw) {
        if (raw == null || raw.isEmpty()) return List.of();
        List<String> out = new ArrayList<>();
        for (String item : raw) {
            String t = trimToNull(item);
            if (t == null) {
                throw new ResponseStatusException(BAD_REQUEST, "No se permiten cantidades menores vacías");
            }
            out.add(t);
        }
        return out;
    }

    static BigDecimal parsePositive(String raw, String field) {
        String n = raw.trim().replace(" ", "").replace(",", ".");
        int i = 0;
        StringBuilder num = new StringBuilder();
        if (i < n.length() && (n.charAt(i) == '+' || n.charAt(i) == '-')) {
            num.append(n.charAt(i));
            i++;
        }
        boolean dot = false;
        while (i < n.length()) {
            char c = n.charAt(i);
            if (c >= '0' && c <= '9') {
                num.append(c);
            } else if (c == '.' && !dot) {
                num.append(c);
                dot = true;
            } else {
                break;
            }
            i++;
        }
        if (num.length() == 0 || num.toString().equals("+") || num.toString().equals("-") || num.toString().equals(".")) {
            throw new ResponseStatusException(BAD_REQUEST, field + " debe ser numérica");
        }
        try {
            BigDecimal v = new BigDecimal(num.toString());
            if (v.compareTo(BigDecimal.ZERO) <= 0) {
                throw new ResponseStatusException(BAD_REQUEST, field + " debe ser mayor a 0");
            }
            return v;
        } catch (NumberFormatException ex) {
            throw new ResponseStatusException(BAD_REQUEST, field + " debe ser numérica");
        }
    }

    static String writeJson(List<String> list) {
        try {
            return JSON.writeValueAsString(list);
        } catch (Exception e) {
            throw new ResponseStatusException(BAD_REQUEST, "No se pudieron guardar las cantidades menores");
        }
    }

    static List<String> readJson(String json) {
        String t = trimToNull(json);
        if (t == null) return List.of();
        try {
            List<String> parsed = JSON.readValue(t, new TypeReference<List<String>>() {});
            if (parsed == null || parsed.isEmpty()) return List.of();
            List<String> out = new ArrayList<>();
            for (String item : parsed) {
                String v = trimToNull(item);
                if (v != null) out.add(v);
            }
            return Collections.unmodifiableList(out);
        } catch (Exception e) {
            String legacy = trimToNull(json);
            return legacy == null ? List.of() : List.of(legacy);
        }
    }

    private static String trimToNull(String v) {
        if (v == null) return null;
        String t = v.trim();
        return t.isEmpty() ? null : t;
    }
}

File: olnatura-qr/web/qr-enterprise-frontend/src/utils/envaseRestos.ts
export const MAX_CANTIDADES_MENORES = 4;

export type LabelCantidades = {
  restosEnabled?: boolean | null;
  envaseTotal?: number | string | null;
  cantidadPorEnvase?: string | null;
  cantidadResto?: string | null;
  restosCantidades?: string[] | null;
} | null | undefined;

export function isRestosEnabled(label: LabelCantidades): boolean {
  return cantidadesMenoresOf(label).length > 0;
}

export function cantidadesMenoresOf(label: LabelCantidades): string[] {
  const list = (label?.restosCantidades ?? [])
    .map((v) => String(v ?? "").trim())
    .filter((v) => v.length > 0);
  if (list.length > 0) return list.slice(0, MAX_CANTIDADES_MENORES);
  if (label?.restosEnabled) {
    const legacy = String(label.cantidadResto ?? "").trim();
    if (legacy) return [legacy];
  }
  return [];
}

export function isCantidadMenorEnvase(label: LabelCantidades, envaseNum: number): boolean {
  const list = cantidadesMenoresOf(label);
  if (list.length === 0) return false;
  const total = Number(label?.envaseTotal);
  if (!Number.isFinite(total)) return false;
  const first = total - list.length + 1;
  return envaseNum >= first && envaseNum <= total;
}

export function cantidadForEnvase(label: LabelCantidades, envaseNum: number): string {
  const list = cantidadesMenoresOf(label);
  if (list.length > 0) {
    const total = Number(label?.envaseTotal);
    const first = total - list.length + 1;
    if (Number.isFinite(total) && envaseNum >= first && envaseNum <= total) {
      return list[envaseNum - first];
    }
  }
  const std = String(label?.cantidadPorEnvase ?? "").trim();
  return std || "N/A";
}

export function parseCantidad(raw: string): number | null {
  const n = String(raw ?? "").trim().replace(/\s/g, "").replace(",", ".");
  const m = n.match(/^[+-]?(\d+(\.\d+)?)/);
  if (!m) return null;
  const v = Number(m[0]);
  if (!Number.isFinite(v) || v <= 0) return null;
  return v;
}

export function cantidadesMenoresOk(
  cantidades: string[],
  cantidadPorEnvase: string,
  envaseTotal: number
): boolean {
  const filled = (cantidades ?? []).map((v) => String(v ?? "").trim()).filter((v) => v.length > 0);
  if (filled.length === 0) return true;
  if (filled.length > MAX_CANTIDADES_MENORES) return false;
  if (envaseTotal < filled.length) return false;
  const std = parseCantidad(cantidadPorEnvase);
  if (std == null) return false;
  return filled.every((raw) => {
    const q = parseCantidad(raw);
    return q != null && q > 0 && q < std;
  });
}

export function envaseQuantities(
  envaseTotal: number,
  cantidadPorEnvase: string,
  cantidadesMenores: string[]
): string[] {
  const total = Math.trunc(Number(envaseTotal));
  if (!Number.isFinite(total) || total < 1) return [];
  const std = String(cantidadPorEnvase ?? "").trim() || "—";
  const menores = (cantidadesMenores ?? [])
    .map((v) => String(v ?? "").trim())
    .filter((v) => v.length > 0)
    .slice(0, Math.min(MAX_CANTIDADES_MENORES, total));
  const firstResto = total - menores.length + 1;
  return Array.from({ length: total }, (_, i) => {
    const envase = i + 1;
    if (menores.length > 0 && envase >= firstResto) {
      return menores[envase - firstResto];
    }
    return std;
  });
}

export function envaseDistribution(
  envaseTotal: number,
  cantidadPorEnvase: string,
  cantidadesMenores: string[]
): string {
  return envaseQuantities(envaseTotal, cantidadPorEnvase, cantidadesMenores)
    .map((q) => `${q} Pz`)
    .join(" / ");
}

export function cantidadTotalOf(label: LabelCantidades): string {
  const total = Number(label?.envaseTotal);
  const qs = envaseQuantities(total, String(label?.cantidadPorEnvase ?? ""), cantidadesMenoresOf(label));
  if (qs.length === 0) return "0";
  let sum = 0;
  let parsed = 0;
  for (const raw of qs) {
    const n = parseCantidad(raw);
    if (n == null) continue;
    sum += n;
    parsed += 1;
  }
  if (parsed === 0) return String(qs.length);
  return Number.isInteger(sum) ? String(sum) : String(sum);
}
